src/datasets/arrow_dataset.py

```diff
@@ -439,6 +439,16 @@ def save_to_disk(self, dataset_path: str, fs=None):
             dataset_path = extract_path_from_uri(dataset_path)
         else:
             fs = fsspec.filesystem("file")
+            cache_files_paths = [Path(cache_filename) for cache_filename in self.cache_files]
+            # Check that the dataset doesn't overwrite iself. It can cause a permission error on Windows and a segfault on linux.
+            if Path(dataset_path).joinpath(DATASET_ARROW_FILENAME) in cache_files_paths:
+                raise PermissionError(
+                    f"Tried to overwrite {Path(dataset_path).joinpath(DATASET_ARROW_FILENAME)} but a dataset can't overwrite itself."
```

Comment:
No it just means that you can't write a Table in the file where it is loaded from (or the Table breaks itself).
i.e. this is forbidden:
```python
path = "path/to/my/dataset"
d = load_from_disk(path)
d = d.rename_column("a", "b")  # this DOESN'T create a new arrow file (the dataset still has the same data)
d.save_to_disk(path)
```

But this is ok:
```python
path = "path/to/my/dataset"
d = load_from_disk(path)
d = d.map(...)  # this DOES create a new arrow file for the new data
d.save_to_disk(path)
```

I'm not 100% convinced by this though. I can probably make this work in all caes by using a temp file rather than writing directly to the arrow file. I'll let you know how this goes.

Reply:
oh got it.  I like map functions creating a new arrow file since we can keep the original dataset unchanged. Thanks a lot.

Reply:
On windows it's not possible to overwrite a file that is already opened for a Table unfortunately so I think we'll have to keep this for now